### PR TITLE
Added support for Huber Loss

### PR DESCRIPTION
**Pull request by alexandrosstergiou**
Current losses in PyTorch only include a (partial) implementation of Huber loss through `smooth l1` based on Fast RCNN - which essentially uses a delta value of 1. Changing/Renaming the [`_smooth_l1_loss()`](https://github.com/pytorch/pytorch/blob/3e1859959a3f720bb5f5e47c3ca15fb3cbfae4da/torch/nn/functional.py#L2487) and refactoring to include delta, enables to use the actual function.

Supplementary to this, I have also made a functional and criterion versions for anyone that wants to set the delta explicitly - based on the functional `smooth_l1_loss()` and the criterion `Smooth_L1_Loss()`

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2511,7 +2511,37 @@ def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='mea
...
+        tens_ops = (input, target)
+        if any([type(t) is not Tensor for t in tens_ops]) and has_torch_function(tens_ops):
+            return handle_torch_function(
+                smooth_l1_loss, tens_ops, input, target, size_average=size_average,
+                reduce=reduce, reduction=reduction)
+    if not (target.size() == input.size()):
+        warnings.warn("Using a target size ({}) that is different to the input size ({}). "
+                      "This will likely lead to incorrect results due to broadcasting. "
+                      "Please ensure they have the same size.".format(target.size(), input.size()),
+                     stacklevel=2)
```
Comment (vincentqb) [inline torch/nn/functional.py:2540]: nit: indentation

File: torch/nn/functional.py [hunk 2]
```diff
@@ -2484,10 +2484,10 @@ def _pointwise_loss(lambd, lambd_optimized, input, target, reduction='mean'):
         return lambd_optimized(expanded_input, expanded_target, _Reduction.get_enum(reduction))
 
 
-def _smooth_l1_loss(input, target):
+def _huber_loss(input, target, delta=1.):
```
Comment (vincentqb) [inline torch/nn/functional.py:2487]: nit: changing the name may break other codes, even though this is a private function

File: torch/nn/functional.py [hunk 3]
```diff
@@ -2511,7 +2511,37 @@ def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='mea
...
+        ret = _huber_loss(input, target)
+        if reduction != 'none':
+            ret = torch.mean(ret) if reduction == 'mean' else torch.sum(ret)
+    else:
+        expanded_input, expanded_target = torch.broadcast_tensors(input, target)
+        ret = torch._C._nn.smooth_l1_loss(expanded_input, expanded_target, _Reduction.get_enum(reduction))
+    return ret
+
+
+def huber_loss(input, target, delta, size_average=None, reduce=None, reduction='mean'):
```
Comment (vincentqb) [inline torch/nn/functional.py:2523]: This looks like the same as `smooth_l1_loss` but for the `delta` parameter. Let's just add a new parameter `delta` with default value `1`.

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -802,6 +801,63 @@ def forward(self, input, target):
         return F.smooth_l1_loss(input, target, reduction=self.reduction)
 
 
+class HuberLoss(_Loss):
```
Comment (vincentqb) [inline torch/nn/modules/loss.py:804]: Again, this looks like the same as `SmoothL1Loss` but for the `delta` parameter. Let's just add `delta` with default value `1.`

**Review by vincentqb — CHANGES_REQUESTED** [review]
I agree with adding the delta parameter, but creating whole new abstractions for this are not necessary.

File: torch/nn/functional.py (hunk 2 above)
Comment (alexandrosstergiou) [reply torch/nn/functional.py]: Changed back to `_smooth_l1_loss`

File: torch/nn/modules/loss.py (hunk 4 above)
Comment (alexandrosstergiou) [reply torch/nn/modules/loss.py]: I agree that programmatically, it would make more sense to simply include the delta value as an additional parameter in both the functional and module versions of Smooth_L1. I have made a new commit based on the comments. Many thanks!

**Review by alexandrosstergiou — COMMENTED** [review]
New changes made based on the comments to include Huber delta as an additional optional parameter in the respective functions. Again, many thanks.

**Review by vincentqb — APPROVED** [review]
LGTM, thanks!

File: torch/nn/functional.py [hunk 5]
```diff
@@ -2582,14 +2582,14 @@ def _pointwise_loss(lambd, lambd_optimized, input, target, reduction='mean'):
         return lambd_optimized(expanded_input, expanded_target, _Reduction.get_enum(reduction))
 
 
-def _smooth_l1_loss(input, target):
-    # type: (Tensor, Tensor) -> Tensor
+def _smooth_l1_loss(input, target, delta=1.):
+    # type: (Tensor, Tensor, float) -> Tensor
     t = torch.abs(input - target)
-    return torch.where(t < 1, 0.5 * t ** 2, t - 0.5)
+    return torch.where(t < delta, 0.5 * t ** 2, t * delta - (0.5 * delta ** 2))
```
Comment (nik-sm) [inline torch/nn/functional.py:2588]: I believe this should be `torch.where(t <= delta, ..., ...)` (see the wiki article for example).

You could consider this thread for example: [https://github.com/devsisters/DQN-tensorflow/issues/16#issuecomment-306162797](https://github.com/devsisters/DQN-tensorflow/issues/16#issuecomment-306162797)

There, someone argues that loss function for training DQN should be Huber with a delta of 2:
> ... to qualify as correct Q-learning, you want to compute an estimate over possible future states... This what you would get with MSE, or Huber loss with behaving as MSE in [-2, 2]

If the implementation of Huber uses `t < delta`, then that "intuitive" delta value of 2 would still not be large enough for DQN.

Comment (vincentqb) [reply torch/nn/functional.py:2588]: When `t=delta`, both expressions are the same, so it wouldn't matter which one is picked. Do I understand your point correctly?

In particular, the first is 
```
0.5 * t ** 2
= 0.5 * delta ** 2.
```
The second is 
```
t * delta - (0.5 * delta ** 2)
= delta * delta - (0.5 * delta ** 2)
= 0.5 * delta ** 2,
```
as before.

File: torch/nn/functional.py [hunk 6]
```diff
@@ -2609,7 +2609,7 @@ def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='mea
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
     if target.requires_grad:
-        ret = _smooth_l1_loss(input, target)
+        ret = _smooth_l1_loss(input, target, delta=delta)
```
Comment (fmassa) [inline torch/nn/functional.py:2612]: You should have also changed lines 2617 (and the corresponding C++ / CUDA implementations).

As it currently stands, the `delta` is not used if `target.requires_grad == False`.

This should have had a test that would have catched this

**Review by fmassa — COMMENTED** [review]
Thanks for the PR, but this is incomplete, and should be reverted in its current state.

For some reference, here is another PR (that got stale) and made the changes I mentioned in the C++ side.

File: torch/nn/functional.py (hunk 5 above)
Comment (nik-sm) [reply torch/nn/functional.py:2588]: Whoops - good point, you're right. (Another way I should have noticed this - there is no discontinuity in the function, so it must be that the two piecewise components have the same value at the border).

Perhaps it would still be more "obviously correct" to have the formula here match the canonical formula, but you're right that there is not a functional difference. Cheers
